src/ucp/core/ucp_worker.c

```diff
@@ -1721,12 +1699,78 @@ static ucs_status_t ucp_worker_init_mpools(ucp_worker_h worker)
         goto err_reg_mp_cleanup;
     }
 
+    for (iface_id = 0; iface_id < worker->num_ifaces; ++iface_id) {
+        if_attr           = &worker->ifaces[iface_id]->attr;
+        max_mp_entry_size = ucs_max(max_mp_entry_size,
+                                    if_attr->cap.am.max_short);
+        max_mp_entry_size = ucs_max(max_mp_entry_size,
+                                    if_attr->cap.am.max_bcopy);
+        max_mp_entry_size = ucs_max(max_mp_entry_size,
+                                    if_attr->cap.am.max_zcopy);
+    }
+
+    /* Create memory pools for incoming UCT messages without a UCT descriptor.
+     * Ignore sizes which are larger than maximal transport segment. */
+    max_idx                = UCP_WORKER_AM_MPS_MAP_SIZE - 1;
+    worker->am_mps_bitmap  = (max_mp_entry_size == 0) ? 0 :
+                             (ucs_roundup_pow2(max_mp_entry_size) - 1) &
+                              context->config.mpool_sizes_map;
+    worker->am_mps_bitmap |= UCS_BIT(max_idx);
+    num_am_mpools          = ucs_popcount(worker->am_mps_bitmap);
+    worker->am_mps         = ucs_malloc(num_am_mpools * sizeof(ucs_mpool_t),
+                                        "am_mpools");
+    if (worker->am_mps == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_rndv_frag_mp_cleanup;
+    }
+
+    mp_index = 0;
+    prev_idx = max_idx;
+    ucs_for_each_bit(size_log2, worker->am_mps_bitmap) {
+        size   = (size_log2 == max_idx) ? max_mp_entry_size :
+                                          UCS_BIT(size_log2);
+        status = ucs_mpool_init(&worker->am_mps[mp_index], 0,
+                                size + UCP_WORKER_HEADROOM_SIZE +
+                                worker->am.alignment,
+                                0, UCS_SYS_CACHE_LINE_SIZE, 128, UINT_MAX,
+                                &ucp_am_mpool_ops, "ucp_am_bufs");
+        if (status != UCS_OK) {
+            goto err_am_mps_cleanup;
+        }
+
+        ucs_assertv(size_log2 < UCP_WORKER_AM_MPS_MAP_SIZE, "size_log2 %d",
+                    size_log2);
+
+        /* am_mps_map is an array of pointers to memory pools. Array index is
+         * log2 of the corresponding memory pool element size. Indexation is
+         * done in the reverse order to speedup array index calculation in
+         * ucp_recv_desc_alloc(). Since there is no separate memory pool for
+         * every pow of 2 value, initialize all array elements in the range of
+         * [current_index, last_initalized_index).
+         * So, eventually every am_mps_map array element will point to a certain
+         * mpool with minimal element size capable of storing elements up to
+         * 2^(max_idx - index). */
+        for (i = prev_idx; i >= max_idx - size_log2; --i) {
+            worker->am_mps_map[i] = &worker->am_mps[mp_index];
+        }
```

Comment:
i think it would be simpler to have another loop after ucs_for_each_bit(..) to populate the missing entries, instead of this nested loop. we have always remember the ptr to last mpool instead of 'prev_idx'